soroban-sdk/src/storage.rs

```diff
@@ -88,56 +112,216 @@ impl Storage {
             Some(V::try_from_val(&self.env, &rv).unwrap_optimized())
         } else {
             None
         }
     }
 
-    /// Sets the value for the given key in the currently executing contract's
-    /// storage.
+    /// Returns the value there is a value stored for the given key in the
+    /// currently executing contract's storage.
     ///
-    /// If the key already has a value associated with it, the old value is
-    /// replaced by the new value.
-    #[inline(always)]
-    pub fn set<K, V>(&self, key: &K, val: &V)
+    /// The returned value is a result of converting the internal value
+    pub(crate) fn set<K, V>(&self, key: &K, val: &V, storage_type: StorageType, flags: Option<u32>)
     where
         K: IntoVal<Env, RawVal>,
         V: IntoVal<Env, RawVal>,
     {
+        let f: RawVal = match flags {
+            None => ().into(),
+            Some(i) => i.into(),
+        };
         let env = &self.env;
-        internal::Env::put_contract_data(env, key.into_val(env), val.into_val(env))
-            .unwrap_infallible();
+        internal::Env::put_contract_data(
+            env,
+            key.into_val(env),
+            val.into_val(env),
+            storage_type,
+            f,
+        )
+        .unwrap_infallible();
+    }
+
+    pub(crate) fn bump<K>(&self, key: &K, storage_type: StorageType, min_ledgers_to_live: u32)
+    where
+        K: IntoVal<Env, RawVal>,
+    {
+        let env = &self.env;
+        internal::Env::bump_contract_data(
+            env,
+            key.into_val(env),
+            storage_type,
+            min_ledgers_to_live.into(),
+        )
+        .unwrap_infallible();
     }
 
     /// Removes the key and the corresponding value from the currently executing
     /// contract's storage.
     ///
     /// No-op if the key does not exist.
     #[inline(always)]
-    pub fn remove<K>(&self, key: &K)
+    pub(crate) fn remove<K>(&self, key: &K, storage_type: StorageType)
     where
         K: IntoVal<Env, RawVal>,
     {
         let env = &self.env;
-        internal::Env::del_contract_data(env, key.into_val(env)).unwrap_infallible();
+        internal::Env::del_contract_data(env, key.into_val(env), storage_type).unwrap_infallible();
     }
 
-    fn has_internal(&self, key: RawVal) -> bool {
-        internal::Env::has_contract_data(&self.env, key)
+    fn has_internal(&self, key: RawVal, storage_type: StorageType) -> bool {
+        internal::Env::has_contract_data(&self.env, key, storage_type)
             .unwrap_infallible()
             .into()
     }
 
-    fn get_internal(&self, key: RawVal) -> RawVal {
-        internal::Env::get_contract_data(&self.env, key).unwrap_infallible()
+    fn get_internal(&self, key: RawVal, storage_type: StorageType) -> RawVal {
+        internal::Env::get_contract_data(&self.env, key, storage_type).unwrap_infallible()
+    }
+}
+
+pub struct Exclusive {
+    storage: Storage,
+}
+
+impl Exclusive {
```

Comment:
nit: I don't think the duplication is necessary here - we could just return differently initialized structs of the same type (so the storage type would just be its field). but this can be cleaned up later, given uncertainty around what types we would have etc.